### DISCUSSION TITLE
Lesson 1: Welcome & Overview

### DISCUSSION THREAD

**Discussion by sampsyo** [sampsyo/cs6120#346] [Lessons]
There's not much to [Lesson 1](https://www.cs.cornell.edu/courses/cs6120/2023fa/lesson/1/)—just write your ideas about what goes into a good _language implementation_ other than optimizations!

**Comment by rcplane** [comment]
A good language implementation should have a written and tested standard specification for all core language features to allow for the possibility of later implementation improvements with less fear of breaking compatible feature behaviors.

This value can be observed in practice from Python interpreter implementations based on C, JVM, attempts to remove GIL, more parallel threading, and type hinting for programmer productivity. ( https://en.wikipedia.org/wiki/Python_(programming_language)  "Major implementations" )

**Comment by HLRJ** [comment]
A good programming language should have a relatively low learning curve.

**Comment by praveer13** [comment]
A good programming language should be simple enough to write without being too bug prone and have safety features built in that protect the programmer from common pitfalls or bugs that can arise out of language semantics or ambiguity of language itself, and maybe enforce good practices out of the box and still be human readable.

**Comment by jiayiyang1997** [comment]
A good programming language should provide more flexibility and allow programmers to touch lower-level features of operating systems. It should also have great readability and have features which allow programmers to apply modern design patterns for software and implement data structures and algorithms conveniently.

**Comment by MadhurKumar004** [comment]
A good programming language implementation should have proper Debuggability attached to it.

**Comment by kumarkshiv** [comment]
A good programming language should have following properties:
- Intuitive: Making it easy and effortless to learn
- Intuitive error messages: Making it easy to debug
- It should make the optimisation effortless for the programmers
- It should exactly capture the intent of the programmer and do the same when executed
- It should allow programmers to validate the expected behaviour during execution (may be by putting some sort of checks)
